### ISSUE TITLE
JSON variable substitute objects and arrays

### ISSUE THREAD

**Issue by stefer** [microsoft/azure-pipelines-tasks#9257]
## Environment
- Server - Azure Pipelines
- Agent - Hosted VS2017

## Issue Description

Task Name: Azure App Service Deploy

We would like the ability to replace arrays (and possibly objects) in JSON variable substitution.

We have application settings with arrays that differ in dev and production environment. It not only differs in content, but also in number of items in the list. There is currently no way to substitute these kind of settings in Azure Devops.

Octopus has support for replacing the whole array with a new array (and also objects), see 
https://octopus.com/docs/deployment-process/configuration-features/json-configuration-variables-feature.

With json file like this:
```
{
   "foo": {
      "bar": [
		{"id": "baz", "name": "qux"},
		{"id": "quux", "name": "corge"},
	  ],
       "garply": {
           "waldo": "fred"
       }
   }
}
```

One could substitute array variable `foo.bar` with value `[{"id: ""uier", "name": "grault"}]`
and possibly object `foo.garply` with value `{"waldo": "plugh", "inserted": "thud"}`

The lack of support for this makes transition from Octopus to Azure Devops hard.

Other people have also asked for this feature:
https://github.com/MicrosoftDocs/vsts-docs/issues/1317

### Task logs
N/A

### Error logs
N/A

**Comment by vincentdass** [comment]
@stefer , thanks for reaching out.
As you know, this would be a breaking change feature in existing task.
We have added a separate File Transform task that does XML Transformation and XML & JSON substitution and accomodated your feature request on the new task.
This should be available in ~5-6 weeks.

**Comment by chr0m1ng** [comment]
Is this working? Docs says that only string can be used... https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/transforms-variable-substitution?view=azure-devops

**Comment by vincentdass** [comment]
@chr0m1ng , apologies the documenation is outdated. We will update it soon

**Comment by chr0m1ng** [comment]
I'm trying to use this feature and isn't working, 
On variables i'm setting the value as the [print](https://imgur.com/DCfeUiE) shows and my appsettings is not updating (when the value is a string it works).
Is this really working?

**Comment by vincentdass** [comment]
@chr0m1ng , Please share the task verison and the debug logs for the same.

**Comment by chr0m1ng** [comment]
> @chr0m1ng , Please share the task verison and the debug logs for the same.

Task version: 4.*

[Debug logs](https://github.com/microsoft/azure-pipelines-tasks/files/3355431/ReleaseLogs_685.zip)

**Comment by chr0m1ng** [comment]
Any news?

**Comment by vincentdass** [comment]
@chr0m1ng , To enable debug logs, add a variable `system.debug` to `true` and rerun the build/release/pipeline.
Also, Is it possible to share the JSON file where you are trying to substitute?

**Comment by fredrikjacobson** [comment]
What Task version for Azure App Service Deploy is required for substituting arrays?
Appears that the File transform task has been updated to support this.

However, using Azure App Service Deploy task version 3.* does not seem to support it.

**Comment by vincentdass** [comment]
@fredrikjacobson ,the feature is available in App Service Deploy task - `4.*` and File Transform - `1.*`

**Comment by chr0m1ng** [comment]
> @chr0m1ng , To enable debug logs, add a variable `system.debug` to `true` and rerun the build/release/pipeline.
> Also, Is it possible to share the JSON file where you are trying to substitute?

Here is, the appsettings.json is the target file

[build_release_logs.zip](https://github.com/microsoft/azure-pipelines-tasks/files/3383318/build_release_logs.zip)

**Comment by vincentdass** [comment]
@chr0m1ng , Apologies , Can you please try the transformation with new task `File Transform` ?
The advanced transformation and substitution options are supported there.

**Comment by ahocquet** [comment]
@vincentdass : I've just tested the new task File Transform, it worked like a charm! Thank you!

**Comment by lekhas01** [comment]
Hi. I am trying to replace a particular key within an array element. i have written the path expression as release variable.But it doesn't work. Can you pls help.
Path Expression(Set as release pipeline variable) : foo.bar[0].id
Sample Json : 
{
   "foo": {
      "bar": [
		{"id": "baz", "name": "qux"},
		{"id": "quux", "name": "corge"}
	  ],
       "garply": {
           "waldo": "fred"
       }
   }
}

**Comment by vincentdass** [comment]
@lekhas01 , please use this variable name:  `foo.bar.0.id`1

**Comment by wolszakp** [comment]
@vincentdass Thanks for releasing advanced transformation. 
I tried that, and it works when I am using: `File Transform` task.
However for `App Service Deploy task` - `4.*` it doesn't work. Do you plan to upgrade `App Service Deploy task` ?

**Comment by baoduy** [comment]
Instead of trying to replace the value in the array with indexing `[0]`. I'm using [Replace Token](https://marketplace.visualstudio.com/items?itemName=qetza.replacetokens&targetId=1a4ff402-8255-461c-bdc2-e455fdf622ff&utm_source=vstsproduct&utm_medium=ExtHubManageList) extension to replace the variable name in the config file with value during deployment.

**Comment by daniellittledev** [comment]
The documentation for this needs to be updated!

**Comment by RoopeshNair** [comment]
@SumiranAgg

**Comment by AngryVirginian** [comment]
I confirm that replacing an entire json array only works in the File Transform task and not in the App Service Deploy task.  Note: this is about replacing the value of the entire array and not replacing value each element via the index position.  E.g., replacing "ids": [1, 2] from the build pipeline with "ids": [1,2,3,4,5] in the release pipeline.

**Comment by codyspeck** [comment]
@AngryVirginian 

I'm curious how you got the File Transform task to replace a json array. This is the final blocker for a feature I am trying to get wrapped up. 

Given the following json file:

`{
  "SomeArray": []
}`

How would you go about setting up your Azure Variables? I've tried setting "SomeArray", but that replaces the array with the string value of the variable. I've also tried setting "SomeArray.0", "SomeArray.1", etc, but that doesn't replace the array at all. I would love for File Transform to be able to do this.

**Comment by AngryVirginian** [comment]
@codyspeck 

What I found working was just replacing the entire content of the entire array.  See below for example.  

![image](https://user-images.githubusercontent.com/28601813/75615370-adb2c500-5b10-11ea-8f22-69283d63cc33.png)

**Comment by muneeb-m** [comment]
@codyspeck, Even I was having problems replacing individual values in a JSON array and stumbled upon this post. I tried out @AngryVirginian's solution and it worked perfectly for me, thanks for that. Though it's a bit annoying seeing my entire JSON array dumped in a single value but anything to get past this silly issue in Azure DevOps.

**Comment by VahidFarahmandian** [comment]
> I confirm that replacing an entire json array only works in the File Transform task and not in the App Service Deploy task. Note: this is about replacing the value of the entire array and not replacing value each element via the index position. E.g., replacing "ids": [1, 2] from the build pipeline with "ids": [1,2,3,4,5] in the release pipeline.

Worked for me. Thank you

**Comment by marcslevesque** [comment]
I was also able to get it working, however:

1. I have a pipeline in which I have chosen IIS Web Deploy (not App Service Deploy)
2. Contrary to the documentation here: https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/transforms-variable-substitution?view=azure-devops&tabs=Classic, I was not able to use the variable substitution from the IIS Web App Deploy task (version 0.*).
3. I had to use the File Transform task (version 1.*) and placed it before the IIS Web App Deploy task.

I also had an array of arrays I needed to replace, since I could not insert new values, and that also worked using the File Transform task.

`        "Variable": [
            [
                "value1",
                "1"
            ],
            [
                "value2",
                "2"
            ],
            [
                "value3",
                "3"
            ]
        ]
 `
